### PR TITLE
Feature Search

### LINKED ISSUE #64: [FR]: Search movie by keyword
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Related app

All - Android

### Describe the problem

As a user, I want to have a search functionality in the movies section or somewhere in the app.

### Describe the solution

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hadi-norouzi**
It's completed but it has some bugs but they can be fixed with other issues if you would agree.

Close #64

### CODE REVIEW

**Comment by moallemi** [comment]
@hadi-norouzi The build is failing.I think instead of merging `dev` to `search` you should rebase `search` with `dev`. I'll fix that.

CleanShot 2024-07-11 at 08 33 07@2x

File: feature/search/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+android {
+  namespace = "io.filmtime.feature.search"
+}
+
+dependencies {
+  implementation(project(":data:model"))
+  implementation(project(":domain:tmdb-search"))
+  implementation(libs.androidx.paging.compose.android)
```
Comment (moallemi) [inline feature/search/build.gradle.kts:13]: We already have `libs.paging.compose`

File: feature/search/src/main/java/io/filmtime/feature/search/SearchViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    _state.update { it.copy(loading = true) }
+
+    viewModelScope.launch {
+      val data = tmdbSearchUseCase(text, type)
+      when (data) {
+        is Failure -> {
+          _state.update { it.copy(error = data.error.toString(), loading = false) }
+        }
+        is Success -> _state.update {
+          it.copy(loading = false, data = data.data, hasResult = data.data.isNotEmpty(), error = null)
```
[the remark this answers is not in the document]
Comment (hadi-norouzi) [reply feature/search/src/main/java/io/filmtime/feature/search/SearchViewModel.kt:34]: I am using Android Studio auto completion for when statement. it do that.

File: data/network/src/main/java/io/filmtime/data/network/TmdbSearchListResponse.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  override val descriptor: SerialDescriptor
+    get() = PolymorphicSerializer(NetworkSearchResult::class).descriptor
+
+  override fun deserialize(decoder: Decoder): NetworkSearchResult {
+    val jsonElement = (decoder as JsonDecoder).decodeJsonElement()
+    return when (val itemType = jsonElement.jsonObject["media_type"]?.jsonPrimitive?.content) {
+      "movie" -> json.decodeFromJsonElement(MovieSearchResult.serializer(), jsonElement)
+      "tv" -> json.decodeFromJsonElement(TvShowSearchResult.serializer(), jsonElement)
+      "person" -> json.decodeFromJsonElement(PersonSearchResult.serializer(), jsonElement)
+      else -> throw SerializationException("Unknown itemType: $itemType")
```
[the remark this answers is not in the document]
Comment (hadi-norouzi) [reply data/network/src/main/java/io/filmtime/data/network/TmdbSearchListResponse.kt:203]: Screenshot_20240711_221834
Is it ok?

**Comment by moallemi** [comment]
Dear @hadi-norouzi. Thanks for your great contribution. 

As you mentioned we have some issues and I think its better to address them before merging this branch. e.g:
- State is lost when you navigate to a search item and press back button. even you can not change between search tabs (All, Move and TV)
- Better handling of that `SerializationException`
- Pagination
- Grid state reset

If you are overwhelmed with this feature, I might have time next weekend and I can contribute to this feature. Let me know WDYT.

**Comment by hadi-norouzi** [comment]
> Dear @hadi-norouzi. Thanks for your great contribution.
> 
> As you mentioned we have some issues and I think its better to address them before merging this branch. e.g:
> 
> * State is lost when you navigate to a search item and press back button. even you can not change between search tabs (All, Move and TV)
> * Better handling of that `SerializationException`
> * Pagination
> * Grid state reset
> 
> If you are overwhelmed with this feature, I might have time next weekend and I can contribute to this feature. Let me know WDYT.

Sure, they are. I have no idea about them but I can implement pagination in this PR and create new issues for others.

**Comment by moallemi** [comment]
You can work on pagination. No need to create new issues. We'll fix them here

**Review by moallemi — APPROVED** [review]

**Pull request merged**